aws/credentials/ibmiam/trusted_profile_credentials_provider.go

```diff
@@ -0,0 +1,171 @@
+package ibmiam
+
+import (
+	"github.com/IBM/go-sdk-core/v5/core"
+	"github.com/IBM/ibm-cos-sdk-go/aws"
+	"github.com/IBM/ibm-cos-sdk-go/aws/awserr"
+	"github.com/IBM/ibm-cos-sdk-go/aws/credentials"
+	"github.com/IBM/ibm-cos-sdk-go/aws/credentials/ibmiam/token"
+)
+
+const (
+	TrustedProfileProviderName = "TrustedProfileProviderNameIBM"
+	tokenType                  = "Bearer"
+)
+
+// TrustedProfileProvider Struct
+// This implements Provider interface
+type TrustedProfileProvider struct {
+	// Name of Provider
+	providerName string
+
+	// Type of Provider - SharedCred, SharedConfig, etc.
+	providerType string
+
+	// Authenticator implements an IAM-based authentication schema
+	authenticator *core.ContainerAuthenticator
+
+	// Error
+	ErrorStatus error
+
+	// Logger attributes
+	logger   aws.Logger
+	logLevel *aws.LogLevelType
+}
+
+// NewTrustedProfileProvider allows the creation of a custom IBM IAM Provider
+// Parameters:
+//
+//	Provider Name
+//	AWS Config
+//	Trusted Profile Name
+//	Trusted Profile ID
+//	Compute Resource Token File Path
+//	IBM IAM Authentication Server Endpoint
+//
+// Returns:
+//
+//	TrustedProfileProvider
+func NewTrustedProfileProvider(providerName string, config *aws.Config, trustedProfileName, trustedProfileID, crTokenFilePath,
+	authEndPoint string) *TrustedProfileProvider {
+	provider := new(TrustedProfileProvider)
+
+	provider.providerName = providerName
+	provider.providerType = "oauth"
+
+	logLevel := aws.LogLevel(aws.LogOff)
+	if config != nil && config.LogLevel != nil && config.Logger != nil {
+		logLevel = config.LogLevel
+		provider.logger = config.Logger
+	}
+	provider.logLevel = logLevel
+
+	if crTokenFilePath == "" {
+		provider.ErrorStatus = awserr.New("crTokenFilePathNotFound", "CR token file path not found", nil)
+		if provider.logLevel.Matches(aws.LogDebug) {
+			provider.logger.Log(debugLog, "<IBM IAM PROVIDER BUILD>", provider.ErrorStatus)
+		}
+
+		return provider
+	}
+
+	if trustedProfileName == "" && trustedProfileID == "" {
+		provider.ErrorStatus = awserr.New("trustedProfileNotFound", "Trusted profile name or id not found", nil)
+		if provider.logLevel.Matches(aws.LogDebug) {
+			provider.logger.Log(debugLog, "<IBM IAM PROVIDER BUILD>", provider.ErrorStatus)
+		}
+
+		return provider
+	}
+
+	if authEndPoint == "" {
+		authEndPoint = defaultAuthEndPoint
+		if provider.logLevel.Matches(aws.LogDebug) {
+			provider.logger.Log(debugLog, "<IBM IAM PROVIDER BUILD>", "using default auth endpoint", authEndPoint)
+		}
+	}
+
+	authenticator, err := core.NewContainerAuthenticatorBuilder().
+		SetIAMProfileName(trustedProfileName).
+		SetIAMProfileID(trustedProfileID).
```

Comment:
What happens if you only set the name, but not the id, as far as I know, it is sufficient to either use the name or ID.

Reply:
Yes it is sufficient to set either ID/TP name. We tested this logic using Unit test case and we were able to authenticate using either one of them

Reply:
You are correct. We added profile id to just give user a choice over name or id. It will work:
- if we set the profile name only
- if we set  the profile id only
- if we set both it should be of the same truster profile, otherwise authentication will fail.